File: mobile-sdk-ios-app/Targets/AppUIKit/Sources/verifier/VerifierHomeView.swift
import SwiftUI

struct VerifierHomeView: View {
    @Binding var path: NavigationPath

    var body: some View {
        VStack {
            VerifierHomeHeader(path: $path)
            VerifierHomeBody(path: $path)
        }
        .navigationBarBackButtonHidden(true)
    }
}

struct VerifierHomeHeader: View {
    @Binding var path: NavigationPath

    var body: some View {
        HStack {
            Text("SpruceKit Demo Verifier")
                .font(.customFont(font: .inter, style: .bold, size: .h2))
                .padding(.leading, 36)
                .foregroundStyle(Color("TextHeader"))
            Spacer()
            Button {
                path.append(VerifierSettingsHome())
            } label: {
                ZStack {
                    RoundedRectangle(cornerRadius: 8)
                        .foregroundColor(Color("Primary"))
                        .frame(width: 36, height: 36)
                    Image("User")
                }
            }
            .padding(.trailing, 20)
        }
        .padding(.top, 10)
    }
}

struct VerifierHomeBody: View {
    @Binding var path: NavigationPath
    
    @State var verificationMethods: [VerificationMethod] = []

    var body: some View {
            ScrollView(.vertical, showsIndicators: false) {
                HStack {
                    Text("VERIFICATIONS")
                        .font(.customFont(font: .inter, style: .bold, size: .p))
                        .foregroundStyle(Color("TextOnPrimary"))
                    Spacer()
                    Text("+ New Verification")
                        .font(.customFont(font: .inter, style: .semiBold, size: .h4))
                        .foregroundStyle(Color("ColorBlue600"))
                        .onTapGesture {
                            // redirect to scanner
                        }
                }

//                VerifierListItem(
//                    title: "Driver's License Document",
//                    description: "Verifies physical driver's licenses issued by the state of Utopia",
//                    binary: true,
//                    fields: 0
//                ).onTapGesture {
//                    path.append(VerifyDL())
//                }

                // Over 21
//                VerifierListItem(
//                    title: "Employment Authorization Document",
//                    description: "Verifies physical Employment Authorization issued by the state of Utopia",
//                    binary: true,
//                    fields: 0
//                ).onTapGesture {
//                    path.append(VerifyEA())
//                }

                // VC
                VerifierListItem(
                    title: "Verifiable Credential",
                    description: "Verifies a Verifiable credential by reading the Verifiable Presentation QR Code",
                    type: VerifierListItemTagType.SCAN_QR_CODE
                ).onTapGesture {
                    path.append(VerifyVC())
                }

                // MDoc
                VerifierListItem(
                    title: "MDoc",
                    description: "Verifies a MDoc by reading the Presentation QR Code",
                    type: VerifierListItemTagType.SCAN_QR_CODE
                ).onTapGesture {
                    path.append(VerifyMDoc())
                }

            }
            .onAppear(perform: {
                self.verificationMethods = VerificationMethodDataStore.shared.getAllVerificationMethods()
            })
            .padding(.all, 24)
    }
}

struct VerifierListItem: View {
    let title: String
    let description: String
    let type: VerifierListItemTagType

    var body: some View {
        VStack(alignment: .leading, spacing: 12) {
            HStack(alignment: .center) {
                Text(title)
                    .font(.customFont(font: .inter, style: .semiBold, size: .h1))
                    .foregroundStyle(Color("TextHeader"))
                Spacer()
                VerifierListItemTag(type: type)
            }
            Text(description)
            Divider()
        }
        .padding(.vertical, 12)
    }
}

enum VerifierListItemTagType {
    case DISPLAY_QR_CODE
    case SCAN_QR_CODE
}

struct VerifierListItemTag: View {
    let type: VerifierListItemTagType

    var body: some View {
        switch type {
        case VerifierListItemTagType.DISPLAY_QR_CODE:
            HStack {
                Image("QRCode")
                    .resizable()
                    .frame(width: 16, height: 16)
                Text("Display")
                    .font(.customFont(font: .inter, style: .semiBold, size: .p))
                    .foregroundStyle(Color.white)
                Image("ArrowTriangleRight")
            }
            .padding(.vertical, 6)
            .padding(.horizontal, 8)
            .background(Color("ColorPurple600"))
            .cornerRadius(100)
        case VerifierListItemTagType.SCAN_QR_CODE:
            HStack {
                Image("QRCodeReader")
                    .resizable()
                    .frame(width: 16, height: 16)
                    .foregroundColor(Color.white)
                Text("Scan")
                    .font(.customFont(font: .inter, style: .semiBold, size: .p))
                    .foregroundStyle(Color.white)
                Image("ArrowTriangleRight")
            }
            .padding(.vertical, 6)
            .padding(.horizontal, 8)
            .background(Color("ColorTerracotta600"))
            .cornerRadius(100)
        }
    }
}

struct VerifierHomeViewPreview: PreviewProvider {
    @State static var path: NavigationPath = .init()

    static var previews: some View {
        VerifierHomeView(path: $path)
    }
}

File: android/Showcase/src/main/java/com/spruceid/mobilesdkexample/verifier/VerifierHomeView.kt
package com.spruceid.mobilesdkexample.verifier

import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Shader
import androidx.compose.ui.graphics.ShaderBrush
import androidx.compose.ui.layout.onSizeChanged
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.IntSize
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.navigation.NavController
import com.spruceid.mobilesdkexample.R
import com.spruceid.mobilesdkexample.navigation.Screen
import com.spruceid.mobilesdkexample.ui.components.HeaderButton
import com.spruceid.mobilesdkexample.ui.components.HomeHeader
import com.spruceid.mobilesdkexample.ui.theme.ColorAmber600
import com.spruceid.mobilesdkexample.ui.theme.ColorBase1
import com.spruceid.mobilesdkexample.ui.theme.ColorBlue600
import com.spruceid.mobilesdkexample.ui.theme.ColorPurple600
import com.spruceid.mobilesdkexample.ui.theme.ColorStone200
import com.spruceid.mobilesdkexample.ui.theme.ColorStone400
import com.spruceid.mobilesdkexample.ui.theme.ColorStone600
import com.spruceid.mobilesdkexample.ui.theme.ColorStone950
import com.spruceid.mobilesdkexample.ui.theme.ColorTerracotta600
import com.spruceid.mobilesdkexample.ui.theme.Switzer
import com.spruceid.mobilesdkexample.utils.activityHiltViewModel
import com.spruceid.mobilesdkexample.viewmodels.VerificationMethodsViewModel

@Composable
fun VerifierHomeView(
    navController: NavController
) {
    Box(
        modifier = Modifier.fillMaxSize()
    ) {
        Column(
            Modifier.fillMaxSize()
        ) {
            VerifierHomeHeader(navController = navController)
            Column(
                Modifier.padding(horizontal = 26.dp)
            ) {
                VerifierHomeBody(
                    navController = navController
                )
            }
        }
    }
}

@Composable
fun VerifierHomeHeader(
    navController: NavController,
) {
    val gradientColors = listOf(ColorAmber600, ColorBase1)
    val buttons = listOf(
        HeaderButton(
            icon = painterResource(id = R.drawable.add),
            contentDescription = stringResource(id = R.string.add),
            onClick = { navController.navigate(Screen.AddVerificationMethodScreen.route) }
        ),
        HeaderButton(
            icon = painterResource(id = R.drawable.cog),
            contentDescription = stringResource(id = R.string.cog),
            onClick = { navController.navigate(Screen.VerifierSettingsHomeScreen.route) }
        )
    )

    HomeHeader(
        title = "Verifier",
        gradientColors = gradientColors,
        buttons = buttons
    )
}

@Composable
fun VerifierHomeBody(
    navController: NavController
) {
    val verificationMethodsViewModel: VerificationMethodsViewModel = activityHiltViewModel()
    val verificationMethods = remember { verificationMethodsViewModel.verificationMethods }

    fun getBadgeType(verificationType: String): VerifierListItemTagType {
        if (verificationType == "DelegatedVerification") {
            return VerifierListItemTagType.DISPLAY_QR_CODE
        } else {
            return VerifierListItemTagType.SCAN_QR_CODE
        }
    }

    LazyColumn(
        Modifier
            .fillMaxWidth()
            .padding(top = 20.dp)
            .padding(bottom = 90.dp)
    ) {

        item {
            VerifierListItem(
                title = "Mobile Driver's License",
                description = "Verifies an ISO formatted mobile driver's license by reading a QR code",
                type = VerifierListItemTagType.SCAN_QR_CODE,
                modifier = Modifier.clickable {
                    navController.navigate(Screen.VerifyMDocScreen.route)
                }
            )
            VerifierListItem(
                title = "Mobile Driver's License - Over 18",
                description = "Verifies an ISO formatted mobile driver's license by reading a QR code",
                type = VerifierListItemTagType.SCAN_QR_CODE,
                modifier = Modifier.clickable {
                    navController.navigate(Screen.VerifyMDlOver18Screen.route)
                }
            )
            VerifierListItem(
                title = "Verify VCB VDL",
                description = "Verify a driver's license encoded as a verifiable credential QRCode",
                type = VerifierListItemTagType.SCAN_QR_CODE,
                modifier = Modifier.clickable {
                    navController.navigate(Screen.VerifyVcbVdlScreen.route)
                }
            )
            VerifierListItem(
                title = "Verifiable Credential",
                description = "Verifies a verifiable credential by reading the verifiable presentation QR code",
                type = VerifierListItemTagType.SCAN_QR_CODE,
                modifier = Modifier.clickable {
                    navController.navigate(Screen.VerifyVCScreen.route)
                }
            )
            VerifierListItem(
                title = "CWT",
                description = "Verifies a CWT by reading a QR code",
                type = VerifierListItemTagType.SCAN_QR_CODE,
                modifier = Modifier.clickable {
                    navController.navigate(Screen.VerifyCWTScreen.route)
                }
            )
            VerifierListItem(
                title = "Driver's License Document",
                description = "Verifies physical driver's licenses issued by the state of Utopia",
                type = VerifierListItemTagType.SCAN_QR_CODE,
                modifier = Modifier.clickable {
                    navController.navigate(Screen.VerifyDLScreen.route)
                }
            )
            VerifierListItem(
                title = "Employment Authorization Document",
                description = "Verifies physical Employment Authorization issued by the state of Utopia",
                type = VerifierListItemTagType.SCAN_QR_CODE,
                modifier = Modifier.clickable {
                    navController.navigate(Screen.VerifyEAScreen.route)
                }
            )
        }
        items(verificationMethods.value) { verificationMethod ->
            VerifierListItem(
                title = verificationMethod.name,
                description = verificationMethod.description,
                type = getBadgeType(verificationMethod.type),
                modifier = Modifier.clickable {
                    navController.navigate(
                        Screen.VerifyDelegatedOid4vpScreen.route.replace(
                            "{id}",
                            verificationMethod.id.toString()
                        )
                    )
                }
            )
        }
    }
}

enum class VerifierListItemTagType {
    DISPLAY_QR_CODE, SCAN_QR_CODE
}

@Composable
fun VerifierListItem(
    title: String,
    description: String,
    type: VerifierListItemTagType,
    modifier: Modifier = Modifier
) {
    Column(
        modifier = modifier.padding(vertical = 12.dp)
    ) {
        Row(
            modifier = Modifier.fillMaxWidth().padding(bottom = 2.dp),
            verticalAlignment = Alignment.CenterVertically
        ) {
            Text(
                text = title,
                fontFamily = Switzer,
                fontWeight = FontWeight.SemiBold,
                fontSize = 16.sp,
                color = ColorStone950,
                modifier = Modifier.weight(4f)
            )
            Spacer(modifier = Modifier.weight(1f))
            VerifierListItemTag(type = type)
        }
        Text(
            text = description,
            fontFamily = Switzer,
            fontWeight = FontWeight.Normal,
            fontSize = 14.sp,
            lineHeight = 19.sp,
            color = ColorStone600,
        )
    }
    HorizontalDivider(color = ColorStone200)
}

@Composable
fun VerifierListItemTag(
    type: VerifierListItemTagType
) {
    when (type) {
        VerifierListItemTagType.DISPLAY_QR_CODE -> {
            Row(
                verticalAlignment = Alignment.CenterVertically,
                modifier = Modifier
                    .clip(shape = RoundedCornerShape(100.dp))
                    .background(ColorPurple600)
                    .padding(vertical = 2.dp)
                    .padding(horizontal = 8.dp),
            ) {
                Image(
                    painter = painterResource(id = R.drawable.qrcode),
                    contentDescription = stringResource(id = R.string.arrow_triangle_right),
                )
                Text(
                    text = "Display",
                    fontFamily = Switzer,
                    fontWeight = FontWeight.Bold,
                    fontSize = 12.sp,
                    color = Color.White,
                    modifier = Modifier.padding(horizontal = 1.dp)
                )
                Image(
                    painter = painterResource(id = R.drawable.arrow_triangle_right),
                    contentDescription = stringResource(id = R.string.arrow_triangle_right),
                )
            }

        }

        VerifierListItemTagType.SCAN_QR_CODE -> {
            Row(
                verticalAlignment = Alignment.CenterVertically,
                modifier = Modifier
                    .clip(shape = RoundedCornerShape(100.dp))
                    .background(ColorTerracotta600)
                    .padding(vertical = 2.dp)
                    .padding(horizontal = 8.dp),
            ) {
                Image(
                    painter = painterResource(id = R.drawable.qrcode_scanner),
                    contentDescription = stringResource(id = R.string.arrow_triangle_right),
                )
                Text(
                    text = "Scan",
                    fontFamily = Switzer,
                    fontWeight = FontWeight.Bold,
                    fontSize = 12.sp,
                    color = Color.White,
                    modifier = Modifier.padding(start = 6.dp, end = 2.dp)
                )
                Image(
                    painter = painterResource(id = R.drawable.arrow_triangle_right),
                    contentDescription = stringResource(id = R.string.arrow_triangle_right),
                )
            }
        }
    }
}
